### ISSUE TITLE
About `max_steps` and chronological prefix sampling in daily data

### ISSUE THREAD

**Issue by Nitasurin** [wangy8989/FinPFN#4]
I would like to ask about the current training sample construction around `max_steps`, especially for daily-frequency data.

From my reading of `scripts/training_utils/data_utils.py`, training samples are generated after sorting dates chronologically:

```python
dates = data_set["date"].drop_duplicates().sort_values().tolist()
date_windows = get_dates_pair(dates, date_style=date_style)

for date_window in date_windows:
    ...
    X_train_list.append(...)
```

Then `DatePairDataset` limits the visible dataset length as:

```python
self.num_steps = min(self.dataset[0].shape[1], max_steps * batch_size)
```

Since `DataLoader(..., shuffle=True, ...)` only shuffles indices within this capped length, my understanding is that when the generated training sample count is larger than `max_steps * batch_size`, training only uses the chronological prefix of the generated samples.

For example, with daily data, `batch_size=64` and `max_steps=100`, this produces only `6400` visible generated samples per epoch. Using the paper's training period of `2016-01-05` to `2021-01-01`, and roughly 500 stocks per date, later date windows can be excluded from training entirely because the cap is applied after chronological sample generation. With `seq_len=100` and two-date windows, each date window contributes about `ceil(500 / 50) = 10` generated samples, so `6400` samples would cover roughly `640` daily date windows, or about 2.5 trading years from the beginning of the training period.

I wanted to ask two questions:

1. Is this prefix-only sampling behavior intentional? In other words, is `max_steps` intended to restrict training to the earliest generated chronological samples, or is this an unintended consequence of where the cap is applied?

2. I tried a local fix where each epoch samples date windows from the full training period before applying the `max_steps` budget, so the selected samples are no longer limited to the early part of the training set. This fixed the coverage issue, but performance degraded severely in my experiments. Do you have any thoughts on why using samples from the full training period could perform worse than the current prefix-only behavior?

I would appreciate your view on whether this should be treated as a deliberate design choice, an implementation issue, or a sampling strategy that needs a different fix.

**Comment by wangy8989** [comment]
`self.num_steps = min(self.dataset[0].shape[1], max_steps * batch_size)`
In this line, `max_steps=100` just sets an upper limit; you can set it `200` or more, it will just slow down. Usually people set it to `100`.
`batch_size=64` just randomly selects `64` date pairs in every step, by `shuffle=True`. So it is not the 2.5 trading years from the beginning (this is when you set `shuffle=False`). 
So if you have `100` draws of `64` date pairs data, it might or might not cover all data, because each draw is random.
But we have more `epochs` to come, so it will cover different combinations in the end, and converge.
I don't know how you did the local fix, but I don't think you need to do so. 
```
    return DataLoader(
        dataset,
        batch_size=batch_size,  # batch size of random indexes draw for every step, not repeated
        shuffle=True,  # True: Good for training to avoid learning patterns based on data(date) order.
        num_workers=num_workers,
        pin_memory=False,
        drop_last=False,  # True: to drop the last incomplete batch
        generator=torch_rng,
        persistent_workers=False,
    )
```

**Comment by Nitasurin** [comment]
Thanks for the reply.

I think the key point is that `shuffle=True` does not change the index range that the `DataLoader` can see.

In the current code:

```python
self.num_steps = min(self.dataset[0].shape[1], max_steps * batch_size)

def __len__(self):
    return self.num_steps
```

So with `max_steps=100` and `batch_size=64`, the dataset length is at most `6400`. `shuffle=True` will shuffle indices inside `range(6400)`, but it will not sample indices after `6400`.

A small fake example shows this:

```python
total_generated_samples = 20
max_steps = 3
batch_size = 4

sample_ids = torch.arange(total_generated_samples, dtype=torch.float32).view(
    1,
    total_generated_samples,
    1,
)
data_tuple = (sample_ids, sample_ids.clone(), sample_ids.clone(), sample_ids.clone())

dataset = DatePairDataset(
    dataset=data_tuple,
    max_steps=max_steps,
    batch_size=batch_size,
)

loader = DataLoader(dataset, batch_size=batch_size, shuffle=True)

seen_ids = []
for batch in loader:
    seen_ids.extend(batch["X_train"].reshape(-1).int().tolist())

print(sorted(seen_ids))
```

The output is a shuffled version of:

```text
[0, 1, 2, ..., 11]
```

It never includes:

```text
[12, 13, ..., 19]
```

So I agree that the exposed samples are shuffled. But the exposed samples are already capped by `__len__()`.

Since `create_data()` builds samples after sorting dates:

```python
dates = data_set["date"].drop_duplicates().sort_values().tolist()
date_windows = get_dates_pair(dates, date_style=date_style)

for date_window in date_windows:
    ...
    X_train_list.append(...)
```

the generated sample order follows the chronological date-window order. Then `DatePairDataset.__len__()` exposes only the first `max_steps * batch_size` generated samples.

So my understanding is:

1. Samples are generated in chronological order.
2. Only the first `max_steps * batch_size` samples are visible to the `DataLoader`.
3. `shuffle=True` shuffles only those visible samples.

That is why I think the current behavior is still prefix-only sampling, just shuffled within the prefix.

If the intended behavior is "100 random batches from all generated date-pair samples", then the cap probably needs to be applied by a sampler, or the date windows need to be sampled before building/truncating the dataset.

**Comment by Nitasurin** [comment]
I tested this with a small fake dataset using the same API shape as the current code:

```python
create_data(...)
DatePairDataset(...)
get_data_loader(...)
```

The fake data has 30 dates and 20 stocks. I added one feature:

```python
marker = date_id * 100 + stock_id
```

So the returned batch tells us both the date window and the stock ids.

For `create_data(data_set=df, seq_len=10, train=False, date_style="consecutive")`, the query-date ids start like this:

```text
[1, 1, 1, 1, 2, 2, 2, 2, ..., 29, 29, 29, 29]
```

So the generated samples follow chronological date-window order. There are many later date windows in the generated data.

Then I rebuilt the training loader through:

```python
get_data_loader(
    train_set=df,
    max_steps=2,
    torch_rng=torch.Generator().manual_seed(seed),
    batch_size=8,
    num_workers=0,
    seq_len=10,
    date_style="consecutive",
)
```

The stock ids can change between rebuilds, because `create_data(..., train=True, ...)` calls:

```python
np.random.choice(stocks, size=(count, stocks_per_date), replace=True)
```

But the visible query-date ids remain:

```text
{1, 2, 3, 4}
```

with:

```python
self.num_steps = min(self.dataset[0].shape[1], max_steps * batch_size)
```

Here `max_steps * batch_size = 16`. With 20 stocks and `seq_len=10`, each date window creates 4 samples, so 16 visible samples cover only the first 4 query dates.

So yes, rebuilding the loader can give different stock combinations. But it still exposes the early date-window prefix after the dataset length is capped.

That is the distinction I meant: the randomness is inside the exposed prefix, not over all generated date-pair samples.

**Comment by wangy8989** [comment]
Yes, you are right. 

I used this because I wanted to make sure it does not produce error when our dataset is smaller than `batch_size*num_steps`:
`self.num_steps = min(self.dataset[0].shape[1], max_steps * batch_size)`
It also ensures that every training epoch has the same number of steps.
But as you said, it exposes the dataloader only to the first `batch_size*num_steps` samples.
So comment out line 200 and use line 199.

Using a sampler can solve this problem.
You can also use a sampler to select 6400 out of 14000 samples every epoch. 

If you use all samples every epoch, you get less randomness; this might be one reason that your performance gets worse.
Or you need to adjust your stopping time or learning rate, etc.
I think the sampler approach can work similarly well.
